`hello/hellospring/service/MemberServiceTest.java`
```java
package hello.hellospring.service;  
  
import hello.hellospring.domain.Member;  
import hello.hellospring.repository.MemoryMemberRepository;  
import org.junit.jupiter.api.AfterEach;  
import org.junit.jupiter.api.BeforeEach;  
import org.junit.jupiter.api.Test;  
  
import static org.assertj.core.api.Assertions.*;  
import static org.junit.jupiter.api.Assertions.*;  
  
class MemberServiceTest {  
    MemberService memberService;  
    MemoryMemberRepository memberRepository;  
  
    @BeforeEach  
    public void beforeEach() {  
        memberRepository = new MemoryMemberRepository();  
        memberService = new MemberService(memberRepository);  
    }  
  
    @AfterEach  
    public void afterEach() {  
        memberRepository.clearStore();  
    }  
  
    @Test  
    public void 회원가입() throws Exception {  
        //Given  
        Member member = new Member();  
        member.setName("hello");  
        //When  
        Long saveId = memberService.join(member);  
        //Then  
        Member findMember = memberRepository.findById(saveId).get();  
        assertEquals(member.getName(), findMember.getName());  
    }  
  
    @Test  
    public void 중복_회원_예외() throws Exception {  
        //Given  
        Member member1 = new Member();  
        member1.setName("spring");  
        Member member2 = new Member();  
        member2.setName("spring");  
        //When  
        memberService.join(member1);  
        IllegalStateException e = assertThrows(IllegalStateException.class,  
                () -> memberService.join(member2));//예외가 발생해야 한다.  
        assertThat(e.getMessage()).isEqualTo("이미 존재하는 회원입니다.");  
    }  
}
```
test에서 메소드명은 한글로도 가능
given  - 주어진 것 / when - 실행한 때 /  then - 결과로 구분하여 작성

예외처리 할 때  try-catch로도 가능하지만 
`assertThrows`를 통해서도 가능

**`assertThrows(Class<> classType, Executable executable)`** [assertThrows](https://velog.io/@cotchan/Exception-%ED%85%8C%EC%8A%A4%ED%8A%B8-%EB%B0%A9%EB%B2%95Junit5)
	: 첫번째 인자로 예상하는 Exception Type을, 
	두 번째 인자로 Executable interface(Exception을 테스트하고자하는 메서드 호출)
	assertThrows의 2번째 인수인 Executable interface method block에서 예상되는 Exception이 발생하면, assertThrows는 해당 Exception을 return value로 반환

DI(Dependency Injection)를 위해 **@beforeEach** 어노테이션을 통해 같은 한 개의 인스턴스를 사용하도록 설정

**@beforeEach** [beforeEach](https://mimah.tistory.com/entry/Spring-Boot-AfterEach-BeforeEach-%EC%98%88%EC%A0%9C)
	: 테스트 메서드 실행 전 무조건 실행

*사용예시)*
 기존 MemberService 클래스에서는 아래와 같이 MemoryMemberRepository 객체를 생성하여 사용한다고 가정
```java
public class MemberService {
    private final MemberRepository memberRepository = new MemoryMemberRepository();    
}
```
MemberService.java에서 DI를 사용해 MemberRepository를 저장
```java
public class MemberService {
    private final MemberRepository memberRepository;
    
    public MemberService(MemberRepository memberRepository) {
        this.memberRepository = memberRepository;
    }
}
```
 테스트 클래스에서 MemberService와 MemoryMemberRepository 객체 생성 부분을 아래와 같이 수정하고, beforeEach() 메서드를 추가

```java
import org.junit.jupiter.api.BeforeEach

class MemberServiceTest {
    MemberService memberService;
    MemoryMemberRepository memberRepository;
    
    @BeforeEach
    public void beforeEach() {
        memberRepository = new MemoryMemberRepository();
        memberService = new MemberService(memberRepository);
    }
}
```